+npxtoolkit/+internal/+logging/log4m.m

```diff
@@ -0,0 +1,270 @@
+            %   PARAMETERS:
+            %       funcName - Name of the function or location from which
+            %       message is coming.
+            %       message - Text of message to log.
+            % 
+            self.writeLog(self.WARN,funcName,message);
+        end
+        
+
+        function error(self, funcName, message)
+            %TRACE Log a message with the ERROR level
+            %
+            %   PARAMETERS:
+            %       funcName - Name of the function or location from which
+            %       message is coming.
+            %       message - Text of message to log.
+            % 
+            self.writeLog(self.ERROR,funcName,message);
+        end
+        
+
+        function fatal(self, funcName, message)
+            %TRACE Log a message with the FATAL level
+            %
+            %   PARAMETERS:
+            %       funcName - Name of the function or location from which
+            %       message is coming.
+            %       message - Text of message to log.
+            % 
+            self.writeLog(self.FATAL,funcName,message);
+        end
+        
+    end
+
+%% Private Methods %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%   Unless you're modifying this, these should be of little concern to you.
+    methods (Access = private)
+        
+        function self = log4m(fullpath_passed)
+            
+            if(nargin > 0)
+                path = fullpath_passed;
+			end
+			self.setFilename(path);
+        end
+        
+%% WriteToFile        
+        function writeLog(self,level,scriptName,message)
+            
+            % If necessary write to command window
+            if( self.commandWindowLevel <= level )
+                fprintf('%s:%s\n', scriptName, message);
+            end
+            
+            %If currently set log level is too high, just skip this log
+            if(self.logLevel > level)
+                return;
+            end 
+            
+            % set up our level string
+            switch level
+                case{self.TRACE}
+                    levelStr = 'TRACE';
+                case{self.DEBUG}
+                    levelStr = 'DEBUG';
+                case{self.INFO}
+                    levelStr = 'INFO';
+                case{self.WARN}
+                    levelStr = 'WARN';
+                case{self.ERROR}
+                    levelStr = 'ERROR';
+                case{self.FATAL}
+                    levelStr = 'FATAL';
+                otherwise
+                    levelStr = 'UNKNOWN';
+            end
+
+            % Append new log to log file
+            try
+                fid = fopen(self.fullpath,'a');
+                fprintf(fid,'%s %s %s - %s\r\n' ...
+                    , datestr(now,'yyyy-mm-dd HH:MM:SS,FFF') ...
+                    , levelStr ...
+                    , scriptName ... % Have left this one with the '.' if it is passed
+                    , message);
+                fclose(fid);
```

Comment:
When calling `fclose` after doing a write, you actually need to test the return value of `fclose` for error status indicators! fclose can in fact fail, because it causes the output buffer to be flushed, which can fail due to out-of-disk-space or other reasons.